job_creator.php

```diff
@@ -0,0 +1,270 @@
+                $matrix['include'][] = $this->createJob(0, [
+                    'phplinting' => true
+                ]);
+            }
+            // phpcoverage also runs unit tests
+            // always run on silverstripe account
+            if ($run['phpcoverage'] || preg_match('#^silverstripe/#', $githubRepository)) {
+                if ($simpleMatrix) {
+                    $matrix['include'][] = $this->createJob(0, [
+                        'phpcoverage' => true
+                    ]);
+                } else {
+                    $matrix['include'][] = $this->createJob(2, [
+                        'db' => DB_MYSQL_57_PDO,
+                        'phpcoverage' => true
+                    ]);
+                }
+            }
+            // endtoend / behat
+            if ($run['endtoend'] && file_exists('behat.yml')) {
+                $matrix['include'][] = $this->createJob(0, [
+                    'endtoend' => true,
+                    'endtoend_suite' => 'root'
+                ]);
+                if (!$simpleMatrix) {
+                    $matrix['include'][] = $this->createJob(3, [
+                        'db' => DB_MYSQL_80,
+                        'endtoend' => true,
+                        'endtoend_suite' => 'root'
+                    ]);
+                }
+            }
+            // javascript tests
+            if (file_exists('package.json') && $run['js']) {
+                $matrix['include'][] = $this->createJob(0, [
+                    'js' => true
+                ]);
+            }
+        }
+
+        // extra jobs
+        foreach ($extraJobs as $arr) {
+            $matrix['include'][] = $this->createJob(0, $arr);
+        }
+
+        // convert everything to strings and sanatise values
+        foreach ($matrix['include'] as $i => $job) {
+            foreach ($job as $key => $val) {
+                if ($val === true) {
+                    $val = 'true';
+                }
+                if ($val === false) {
+                    $val = 'false';
+                }
+                // all values must be strings
+                $val = (string) $val;
+                // remove any dodgy characters
+                $val = str_replace(["\r", "\n", "\t", "'", '"', '&', '|'], '', $val);
+                // ascii chars only - https://www.regular-expressions.info/posixbrackets.html
+                $val = preg_replace('#[^\x20-\x7E]#', '', $val);
+                // limit name_suffix length and be strict as it is used in the artifact file name
+                if ($key === 'name_suffix') {
+                    if (strlen($val) > 20) {
+                        $val = preg_replace('#[^a-zA-Z0-9_\- ]#', '', $val);
+                        $val = substr($val, 0, 20);
+                    }
+                }
+                // be strict with composer_require_extra
```

Comment:
"be strict" in this sense meaning what? What are we trying to fix or avoid with this regex? Where did the list of allowed characters come from? If these are package names and constraints, it might be better to reject values (and maybe throw an exception) which have disallowed characters instead of trying to 'massage' them to become valid.

Reply:
Change to throw an exception